client/README.md

```diff
@@ -0,0 +1,233 @@
+
+---
+
+### **1. `verifyResourceAccess`**
+**Checks if the current user has access to a resource** based on predefined **scopes**.
+
+#### **Method Signature:**
+```java
+void verifyResourceAccess(String resourceId, String resourceIndex, Set<String> scopes, ActionListener<Boolean> listener);
+```
+
+#### **Example Usage:**
+```java
+Set<String> scopes = Set.of("READ_ONLY");
+resourceSharingClient.verifyResourceAccess(
+    "resource-123",
+    "resource_index",
+    scopes,
+    ActionListener.wrap(isAuthorized -> {
+        if (isAuthorized) {
+            System.out.println("User has access to the resource.");
+        } else {
+            System.out.println("Access denied.");
+        }
+    }, e -> {
+        System.err.println("Failed to verify access: " + e.getMessage());
+    })
+);
+```
+> **Use Case:** Before performing operations like **deletion or modifications**, ensure the user has the right permissions.
+
+---
+
+### **2. `shareResource`**
+**Grants access to a resource** for specific users, roles, or backend roles.
+
+#### **Method Signature:**
+```java
+void shareResource(String resourceId, String resourceIndex, Map<String, Object> shareWith, ActionListener<ResourceSharing> listener);
+```
+
+#### **Example Usage:**
+```java
+Map<String, Object> shareWith = Map.of(
+    "users", List.of("user_1", "user_2"),
+    "roles", List.of("admin_role"),
+    "backend_roles", List.of("backend_group")
+);
+
+resourceSharingClient.shareResource(
+    "resource-123",
+    "resource_index",
+    shareWith,
+    ActionListener.wrap(response -> {
+        System.out.println("Resource successfully shared with: " + shareWith);
+    }, e -> {
+        System.err.println("Failed to share resource: " + e.getMessage());
+    })
+);
+```
+> **Use Case:** Used when an **owner/admin wants to share a resource** with specific users or groups.
+
+---
+
+### **3. `revokeResourceAccess`**
+**Removes access permissions** for specified users, roles, or backend roles.
+
+#### **Method Signature:**
+```java
+void revokeResourceAccess(String resourceId, String resourceIndex, Map<String, Object> entitiesToRevoke, Set<String> scopes, ActionListener<ResourceSharing> listener);
+```
+
+#### **Example Usage:**
+```java
+Map<String, Object> entitiesToRevoke = Map.of(
+    "users", List.of("user_2"),
+    "roles", List.of("viewer_role")
+);
+Set<String> scopesToRevoke = Set.of("READ_ONLY");
+
+resourceSharingClient.revokeResourceAccess(
+    "resource-123",
+    "resource_index",
+    entitiesToRevoke,
+    scopesToRevoke,
+    ActionListener.wrap(response -> {
+        System.out.println("Resource access successfully revoked for: " + entitiesToRevoke);
+    }, e -> {
+        System.err.println("Failed to revoke access: " + e.getMessage());
+    })
+);
+```
+> **Use Case:** When a user no longer needs access to a **resource**, their permissions can be revoked.
+
+---
+
+### **4. `listAllAccessibleResources`**
```

Comment:
I know there's been a bit of back-and-forth on design, but I think we should avoid methods like listAll and use standard search APIs. Previously, I know we discussed a DLS-Style "clause injection" on top of any search query that a plugin performs on a resource index and I think something like that would cover any type of search request that a plugin would perform.